code/game/machinery/telecomms/machines/message_server.dm

```diff
@@ -0,0 +1,193 @@
+	var/newKey
+	newKey += pick("the", "if", "of", "as", "in", "a", "you", "from", "to", "an", "too", "little", "snow", "dead", "drunk", "rosebud", "duck", "al", "le")
+	newKey += pick("diamond", "beer", "mushroom", "assistant", "clown", "captain", "twinkie", "security", "nuke", "small", "big", "escape", "yellow", "gloves", "monkey", "engine", "nuclear", "ai")
+	newKey += pick("1", "2", "3", "4", "5", "6", "7", "8", "9", "0")
+	return newKey
+
+/obj/machinery/telecomms/message_server/process()
+	if(toggled && (stat & (BROKEN|NOPOWER)))
+		toggled = FALSE
+	update_icon()
+
+/obj/machinery/telecomms/message_server/receive_information(datum/signal/subspace/pda/signal, obj/machinery/telecomms/machine_from)
+	// can't log non-PDA signals
+	if(!istype(signal) || !signal.data["message"] || !toggled)
+		return
+
+	// log the signal
+	pda_msgs += new /datum/data_pda_msg(signal.format_target(), "[signal.data["name"]] ([signal.data["job"]])", signal.data["message"], signal.data["photo"])
+	signal.data -= "reject"  // only gets through if it's logged
+
+	// pass it along to either the hub or the broadcaster
+	if(!relay_information(signal, /obj/machinery/telecomms/hub))
+		relay_information(signal, /obj/machinery/telecomms/broadcaster)
+
+/obj/machinery/telecomms/message_server/update_icon()
+	if((stat & (BROKEN|NOPOWER)))
+		icon_state = "server-nopower"
+	else if (!toggled)
+		icon_state = "server-off"
+	else
+		icon_state = "server-on"
+
+
+// Repath for maps
+/obj/machinery/message_server
+	parent_type = /obj/machinery/telecomms/message_server
+
+
+// PDA signal datum
+/datum/signal/subspace/pda
+	frequency = FREQ_COMMON
+	server_type = /obj/machinery/telecomms/message_server
+
+/datum/signal/subspace/pda/New(source, data)
+	src.source = source
+	src.data = data
+	var/turf/T = get_turf(source)
+	levels = list(T.z)
+	data["reject"] = TRUE  // set to FALSE if a messaging server logs it
+
+/datum/signal/subspace/pda/copy()
+	var/datum/signal/subspace/pda/copy = new(source, data.Copy())
+	copy.original = src
+	copy.levels = levels
+	return copy
+
+/datum/signal/subspace/pda/proc/format_target()
+	if (length(data["targets"]) > 1)
+		return "Everyone"
+	return data["targets"][1]
+
+/datum/signal/subspace/pda/proc/format_message()
+	if (data["photo"])
+		return "[data["message"]] <a href='byond://?src=[REF(src)];photo=1'>(Photo)</a>"
+	return data["message"]
+
+/datum/signal/subspace/pda/broadcast()
+	for (var/obj/item/device/pda/P in GLOB.PDAs)
+		if ("[P.owner] ([P.ownjob])" in data["targets"])
+			P.receive_message(src)
+
+/datum/signal/subspace/pda/Topic(href, href_list)
+	..()
+	if (href_list["photo"])
+		var/mob/M = usr
+		M << browse_rsc(data["photo"], "pda_photo.png")
+		M << browse("<html><head><title>PDA Photo</title></head>" \
+		+ "<body style='overflow:hidden;margin:0;text-align:center'>" \
+		+ "<img src='pda_photo.png' width='192' style='-ms-interpolation-mode:nearest-neighbor' />" \
+		+ "</body></html>", "window=book;size=192x192")

```

Comment:
`window=pdaphoto` and `onclose(M,"pdaphoto")` or similar. (This is probably my copypaste 🙈 )